devito/types/dense.py

```diff
@@ -49,49 +51,48 @@ class DiscreteFunction(AbstractCachedFunction, ArgProvider):
     is_DiscreteFunction = True
     is_Tensor = True
 
-    def __init__(self, *args, **kwargs):
-        if not self._cached():
-            # A `Distributor` to handle domain decomposition (only relevant for MPI)
-            self._distributor = self.__distributor_setup__(**kwargs)
-
-            # Staggering metadata
-            self._staggered, self.is_Staggered = self.__staggered_setup__(**kwargs)
-
-            # Now that *all* __X_setup__ hooks have been called, we can let the
-            # superclass constructor do its job
-            super(DiscreteFunction, self).__init__(*args, **kwargs)
-
-            # There may or may not be a `Grid` attached to the DiscreteFunction
-            self._grid = kwargs.get('grid')
-
-            # Symbolic (finite difference) coefficients
-            self._coefficients = kwargs.get('coefficients', 'standard')
-            if self._coefficients not in ('standard', 'symbolic'):
-                raise ValueError("coefficients must be `standard` or `symbolic`")
-
-            # Data-related properties and data initialization
-            self._data = None
-            self._first_touch = kwargs.get('first_touch', configuration['first-touch'])
-            self._allocator = kwargs.get('allocator', default_allocator())
-            initializer = kwargs.get('initializer')
-            if initializer is None or callable(initializer):
-                # Initialization postponed until the first access to .data
-                self._initializer = initializer
-            elif isinstance(initializer, (np.ndarray, list, tuple)):
-                # Allocate memory and initialize it. Note that we do *not* hold
-                # a reference to the user-provided buffer
-                self._initializer = None
-                if len(initializer) > 0:
-                    self.data_with_halo[:] = initializer
-                else:
-                    # This is a corner case -- we might get here, for example, when
-                    # running with MPI and some processes get 0-size arrays after
-                    # domain decomposition. We touch the data anyway to avoid the
-                    # case ``self._data is None``
-                    self.data
+    def __init_finalize__(self, *args, **kwargs):
```

Comment:
note for reviewers: this is exactly like before, but de-indented as we don't need ``if not self._cached()`` anymore